### PR TITLE
objc: add safe Objc Classes

### LINKED ISSUE #143: enable `RegisterClass` for a struct with member variables
Now `RegisterClass` is available only for structs without member variables. If a struct had member variables, memory corruption might happen (`examples/objc` uses `RegisterClass` for a struct with a member variable, but this is a known issue).

As @TotallyGamerJet and Hajime discussed on Ebitengine Discord, would it be possible giving up mirroring member variables in the Go world to the Objective-C world? This makes the Objective-C struct size predictable and then memory corruption would no longer happen. Unfortunately, this would be a breaking change since `GetIvar` and `SetIvar` no longer works for Go member variables. However, the current major version is still 0, Hajime is fine to have such a breaking change.

Currently, a struct with members can use `AllocateClassPair`, `AddMethod`, and `AddIvar`, but this is quite painful. `RegisterClass` cares all of them, which is pretty nice.

### PR DESCRIPTION
**Pull request by TotallyGamerJet**
(now actually safe)

The issued lied in two areas. One, for loop variables take the address instead of copying the memory and taking that address. This should be fixed in 1.22. The second area of issues was caused by memory corruption occurring due to calling C functions inside of closures. 

I've tested this PR with ebitengine and all tested passed on amd64 & arm64 darwin.

Closes #143

### CODE REVIEW

File: objc/objc_runtime_darwin.go [hunk 1]
```diff
@@ -148,203 +170,165 @@ func AllocateClassPair(super Class, name string, extraBytes uintptr) Class {
...
+// Attribute determines if a getter and or setter method is generated for this field.
+type FieldDef struct {
+	Name      string
+	Type      reflect.Type
+	Attribute IvarAttrib
+}
 
-		skipSpace()
+// ivarRegex checks to make sure the Ivar is correctly formatted
+var ivarRegex = regexp.MustCompile("[a-z][a-zA-Z0-9]*")
```
Comment (hajimehoshi) [inline objc/objc_runtime_darwin.go:213]: I realized a name can incude `_`. So

`[a-z_][a-zA-Z0-9_]`

is more exact?

https://google.github.io/styleguide/objcguide.html#category-naming

File: objc/objc_runtime_darwin.go [hunk 2]
```diff
@@ -148,203 +170,165 @@ func AllocateClassPair(super Class, name string, extraBytes uintptr) Class {
...
+				nil, false,
+			)
+			var encoding string
+			if encoding, err = encodeFunc(reflect.New(ty).Elem().Interface()); err != nil {
+				return 0, fmt.Errorf("objc: failed to create read method for '%s': %w", ivar.Name, err)
+			}
+			val := reflect.MakeFunc(ty, func(args []reflect.Value) (results []reflect.Value) {
+				// on entry the first and second arguments are ID and SEL followed by the value
+				if len(args) != 3 {
+					panic(fmt.Errorf("objc: incorrect number of arguments"))
```
Comment (hajimehoshi) [inline objc/objc_runtime_darwin.go:281]: This doesn't have to an error, so `Sprintf` is fine.

Give the actual number of arguments and the expected number of arguments to Errorf to make the panic message rich.

File: objc/objc_runtime_darwin.go [hunk 3]
```diff
@@ -148,203 +170,165 @@ func AllocateClassPair(super Class, name string, extraBytes uintptr) Class {
...
+				[]reflect.Type{ivar.Type}, false,
+			)
+			var encoding string
+			if encoding, err = encodeFunc(reflect.New(ty).Elem().Interface()); err != nil {
+				return 0, fmt.Errorf("objc: failed to create read method for '%s': %w", ivar.Name, err)
+			}
+			val := reflect.MakeFunc(ty, func(args []reflect.Value) (results []reflect.Value) {
+				// on entry the first and second arguments are ID and SEL
+				if len(args) != 2 {
+					panic(fmt.Errorf("objc: incorrect number of arguments"))
```
Comment (hajimehoshi) [inline objc/objc_runtime_darwin.go:315]: ditto

**Comment by hajimehoshi** [comment]
I'll take a look again tomorrow. Good night!

File: objc/objc_runtime_darwin.go [hunk 4]
```diff
@@ -148,203 +170,165 @@ func AllocateClassPair(super Class, name string, extraBytes uintptr) Class {
...
 		}
 	}
 	// Add Ivars
-	// Start at 1 because we skip the class object which is first
-	for i := 1; i < strct.NumField(); i++ {
-		f := strct.Field(i)
-		if f.Name == "_" {
-			continue
+	for _, instVar := range ivars {
+		ivar := instVar
```
[the remark this answers is not in the document]
Comment (TotallyGamerJet) [reply objc/objc_runtime_darwin.go:252]: Shadowing makes my IDE get mad so I try to avoid doing it.

[the remark this answers is not in the document]
Comment (hajimehoshi) [reply objc/objc_runtime_darwin.go:252]: Hmm, wouldn't it be possible the IDE side?

**Review by hajimehoshi — APPROVED** [review]
LGTM (the variable name issue is optional)

File: objc/objc_runtime_darwin.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (TotallyGamerJet) [reply objc/objc_runtime_darwin.go:252]: I could change it but I find it helpful to spot bugs. I know you said it's optional so I'd like to keep it the way it is if that's ok.

**Pull request merged**